src/core/player.cpp

```diff
@@ -213,7 +215,25 @@ void Player::NextItem(Engine::TrackChangeFlags change) {
   // Manual track changes override "Repeat track"
   const bool ignore_repeat_track = change & Engine::Manual;
 
-  int i = active_playlist->next_row(ignore_repeat_track);
+  int i = active_playlist->current_row();
+  int original_index = i;
+  if (next_album && i != -1) {
+    if (active_playlist->sequence()->repeat_mode() !=
+        PlaylistSequence::Repeat_Track) {
```

Comment:
If you decide that it should be a no-op, then we should probably grey out the menu item.

There's also a Repeat_Album mode. I think it would loop back to your starting point there as well. I haven't looked at the behavior of the other repeat modes.

Reply:
Here's the list of modes and their respective behaviors when there is only one album in the playlist: 
- Don't Repeat: Stops when it finds there are no further items to play;
- Repeat Track: Replays the same track;
- Repeat Album: Replays the same track;
- Repeat Playlist: Replays the same track;
- Stop after each Track: Stops immediately (same behavior as next track in this case)
- Intro Tracks: Stops when it finds there are no further items to play.

I'm not 100% sure this is exactly the expected behavior, but then again doing a "next album" on a playlist with only one album is a bit of an edge case.

As far as greying out the menu item is concerned, I'd have to look at tying that in with the selection of the "Repeat Track" option.

Reply:
In the painfully long `MainWindow::PlaylistRightClick` method, we have access to the current playlist and we already call `setEnabled` on quite a few items, so that should be pretty straight forward.

Reply:
I think Repeat Album would give you the same behavior if you had multiple albums in the playlist.

Reply:
Examining the code it looks like the selection of the Repeat Mode is handled in the PlaylistSequence object. It is not clear to me how this action can be tied back to changing the state of a component on the MainWindow. Do you have any suggestions?

And yes you're right - when Repeat Mode is Repeat Album, the behavior is the also same when there are multiple albums in the playlist.

Reply:
In PlaylistRightClick, you should be able to access  `app_->playlist_manager()->current()->sequence()`. It looks like this is the same object as ui_->playlist_sequence, but it's probably best not to make that assumption. In fact, it would be nice if Playlist had a GetRepeatMode() accessor.

Reply:
I understand how to get to the PlaylistSequence from MainWindow, however what I am struggling with is that I'd need access to the MainWindow from PlaylistSequence and I don't have a good solution for that.

Changing the repeat mode triggers the following in PlaylistSequence:

void PlaylistSequence::RepeatActionTriggered(QAction* action) {}

I am assuming that a change to the repeat mode, where repeat mode becomes Repeat_Track, an update to the MainWindow needs to be made to deselect the next album option. Maybe I'm not looking at this right but I don't see or haven't found an entry point in MainWindow for a change of repeat mode?

Reply:
Ignore my previous post. I figured out how to handle the "repeat album" signal in MainWindow.cpp and have added a commit that disables the "next album" option when repeat mode is set (or changed) to "repeat track". I wonder if I shoud expand that to include "repeat album" as well?

Reply:
You shouldn't need to handle the signal. In other cases, we enable/disable items when we show the menu. Again, take a look at PlaylistRightClick.

Reply:
Regarding the repeat album mode, I think you would want to disable the menu item for that as well. But when/if we address that other issue that I mentioned, we probably want to follow the same behavior here. Both would require some plumbing in the playlist, though.

Reply:
> You shouldn't need to handle the signal. In other cases, we enable/disable items when we show the menu. Again, take a look at PlaylistRightClick.

I was of the impression that PlaylistRightClick shows a number of good examples of how to enable/disable items on the UI given specific circumstances. I tried to recreate that behavior in the new "SetNextAlbumEnabled" method. What is not clear to me is how I find out the Repeat Mode menu was accessed other than through a specific signal? This would then force a change of the availability of the Next Album function.

Reply:
> Regarding the repeat album mode, I think you would want to disable the menu item for that as well. But when/if we address that other issue that I mentioned, we probably want to follow the same behavior here. Both would require some plumbing in the playlist, though.

I get what you're saying - in Repeat Album mode, the Next Album action should still cause the next album to be selected. Agreed this would involve some more plumbing work, maybe for a future pull request? I tried to recreate #6860 without success so at this point I'm not sure what to do about that one...

Reply:
> I get what you're saying - in Repeat Album mode, the Next Album action should still cause the next album to be selected. Agreed this would involve some more plumbing work, maybe for a future pull request? I tried to recreate #6860 without success so at this point I'm not sure what to do about that one...

Yeah, I wasn't suggesting that it should be fixed with this change.

Reply:
> I was of the impression that PlaylistRightClick shows a number of good examples of how to enable/disable items on the UI given specific circumstances. I tried to recreate that behavior in the new "SetNextAlbumEnabled" method. What is not clear to me is how I find out the Repeat Mode menu was accessed other than through a specific signal? This would then force a change of the availability of the Next Album function.

It looks like we do set menu items in slots for the main controls. But in those cases, we also need to update main controls that are always visible.

A potential issue that I see with your new approach are that it assumes that the repeat behavior is global. In the future, if you broke that assumption, you would also have to handle the update when you changed current playlists. It could work, it just seems more complex.

Reply:
Not sure I understand the question about the repeat mode menu. We're talking about the playlist menu, right?

Reply:
Now I think I understand where the confusion lies. I changed the behavior of the Next Track button to have a pulldown menu (like the Stop button):

![Screenshot_2021-01-19_09-06-00](https://user-images.githubusercontent.com/18431242/105050954-4eb5b180-5a3c-11eb-9cc6-3b1ad0dbf36d.png)

There is a submenu from that button:

![screen_grab](https://user-images.githubusercontent.com/18431242/105051049-642adb80-5a3c-11eb-9460-8fb6dddbef3c.jpg)

So the action is not tied to the right-click menu but a new menu on the UI - hence the need to tie the functionality to a specific signal,

Reply:
Oh, I see. Sorry, I missed that. That makes more sense. I think you could also use the `QMenu::aboutToShow` signal if you wanted to update on show.

Reply:
The RepeatModeChanged signal was already used in MainWindow.cpp so I tied the behavior to that.

Reply:
Looks good to me. You probably want to squash some of the commits.

Reply:
I thought I squashed the commits so all the changes for this would be on a single commit but somehow it appears to have pulled in a bunch of other changes. It didn't change the number of files affected by this Pull Request though, so I guess it's ok? (Still a bit of a noob at git)

Reply:
You definitely have 24 separate commits in this set.
If you haven't already, set up a remote for this repo. Example: `git remote add upstream git@github.com:clementine-player/Clementine.git` then `git fetch upstream`
You can see how many commits your are ahead of the upstream master with `git cherry -v upstream/master`
Rebase: `git rebase -i upstream/master`
Now if you run `git -v cherry upstream/master` again, you should just see the commits that haven't been squashed and `git diff origin/master` should show no code differences.
Now you can force push your branch.

You might want to branch a backup before you do this in case you make a mistake. Commits can be recovered, but it's easier if you have a reference. Also, in the future, you might want to work off of a task branch instead of your master so you can keep your master up to date with the the upstream.

Reply:
Also, hatstand had a comment that you didn't address. And he's the one that pushes the merge button.

Reply:
Thanks for the hints! I ran through the commands you suggested and managed to properly squash the commits into one. Now the only thing left to do is to address hatstand's comment which I'd completely forgotten about...